docs/zkapps/o1js/bitwise-gadgets.mdx

```diff
@@ -0,0 +1,320 @@
+---
+title: Bitwise Gadgets
+hide_title: true
+description: Bitwise gadgets in o1js.
+keywords:
+  - gadgets
+  - bitwise
+  - o1js
+---
+
+:::info
+
+Please note that zkApp programmability is not yet available on Mina Mainnet, but zkApps can now be deployed to Berkeley Testnet.
+
+:::
+
+# Bitwise Gadgets in o1js
+
+Bitwise operations work on 2-bit patterns of equal lengths by positionally matching their individual bits.
+
+- [and](#and)
+- [compactMultiRangeCheck](#compactmultirangecheck)
+- [leftShift](#leftshift)
+- [multiRangeCheck](#multirangecheck)
+- [not](#not)
+- [rangeCheck64](#rangecheck64)
+- [rightShift](#rightshift)
+- [rotate](#rotate)
+- [xor](#xor)
+
+## and
+
+```ts
+and(a: Field, b: Field, length: number) {
+    return and(a, b, length);
+```
+
+The bitwise and gadget on Field elements is equivalent to the [bitwise AND (&)](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Bitwise_AND) operator in JavaScript. 
+
+- The and gadget works by comparing two bits and returns `1` only if both bits are `1`, otherwise it returns `0`.
+- Constrains both input values to fit into `paddedLength` bits. 
+- Both Field elements must fit into `2^paddedLength - 1` or an error is thrown.
+- Guarantees that the output is has at most `paddedLength` bits.
+- Can be checked by a double generic gate that verifies the following relationship between the values. In the process, it also invokes the [xor()](#xor) gadget that creates additional constraints depending on `length`.
+
+  The generic gate verifies `a + b = sum` and the conjunction equation `2 * and = sum - xor`.
+
+  where:
+  * `a + b = sum`
+  * `a ^ b = xor`
+  * `a & b = and`
+
+For details about the implementation, see [AND](https://o1-labs.github.io/proof-systems/specs/kimchi.html?highlight=gates#and) in the Mina book.
+
+The `length` parameter:
+
+- Defines how many bits to compare. 
+- Rounds `length` to the nearest multiple of 16, `paddedLength = ceil(length / 16) * 16`.
+- Specifying a larger `length` parameter adds additional constraints.
```

Comment:
Two questions here:
1. "adds additional" is redundant. Wdyt about "adds more constraints" or just "adds constraints" (but what are the constraints?)

2. does this consideration on https://github.com/o1-labs/docs2/pull/737/files#diff-c8519360c4a9410ca73564931a0167afc43ba115e125b3719e1c6fa8edc2adc3R305 apply to all gadgets? tradeoffs?